File: app/src/main/java/ru/rsppv/criminalintent/fragment/CrimeListFragment.java
package ru.rsppv.criminalintent.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

import ru.rsppv.criminalintent.CrimePagerActivity;
import ru.rsppv.criminalintent.R;
import ru.rsppv.criminalintent.model.Crime;
import ru.rsppv.criminalintent.model.CrimeLab;

public class CrimeListFragment extends Fragment {
    private static final String DATE_PATTERN = "EEE, d MMM yyyy";
    private RecyclerView mCrimeRecyclerView;
    private CrimeAdapter mAdapter;


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_crime_list, container, false);
        mCrimeRecyclerView = (RecyclerView) view.findViewById(R.id.crime_recycler_view);
        mCrimeRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

        updateUI();

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        updateUI();
    }

    private void updateUI() {
        if (mAdapter == null) {
            CrimeLab crimeLab = CrimeLab.get(getActivity());
            mAdapter = new CrimeAdapter(crimeLab.getCrimes());
            mCrimeRecyclerView.setAdapter(mAdapter);
        } else {
            mAdapter.notifyDataSetChanged();
        }
    }

    private class CrimeHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private TextView mTitleTextView;
        private TextView mDateTextView;
        private ImageView mCrimeSolvedImageView;
        private Crime mCrime;

        public CrimeHolder(LayoutInflater inflater, ViewGroup parent) {
            super(inflater.inflate(R.layout.list_item_crime, parent, false));
            itemView.setOnClickListener(this);

            mTitleTextView = (TextView) itemView.findViewById(R.id.crime_title);
            mDateTextView = (TextView) itemView.findViewById(R.id.crime_date);
            mCrimeSolvedImageView = (ImageView) itemView.findViewById(R.id.crime_solved);
        }

        private void bind(Crime crime) {
            mCrime = crime;
            mTitleTextView.setText(crime.getTitle());
            mDateTextView.setText(formatDate(crime));
            mCrimeSolvedImageView.setVisibility(crime.isSolved() ? View.VISIBLE : View.GONE);
        }

        private String formatDate(Crime crime) {
            SimpleDateFormat format = new SimpleDateFormat(DATE_PATTERN);
            return format.format(crime.getDate());
        }

        @Override
        public void onClick(View v) {
            Intent intent = CrimePagerActivity.createIntent(getActivity(), mCrime.getId());
            startActivity(intent);
        }
    }


    private class CrimeAdapter extends RecyclerView.Adapter<CrimeHolder> {
        private List<Crime> mCrimes;

        public CrimeAdapter(Collection<Crime> crimes) {
            mCrimes = new ArrayList<>(crimes);
        }

        @Override
        public CrimeHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(getActivity());
            return new CrimeHolder(inflater, parent);
        }

        @Override
        public void onBindViewHolder(CrimeHolder holder, int position) {
            holder.bind(mCrimes.get(position));
        }

        public void notifyCrimeChanged(UUID crimeId) {
            Crime crime = CrimeLab.get(getActivity()).getCrime(crimeId);
            notifyItemChanged(mCrimes.indexOf(crime));
        }

        @Override
        public int getItemCount() {
            return mCrimes.size();
        }
    }
}

File: app/src/main/java/ru/rsppv/criminalintent/fragment/CrimeListFragment.kt
package ru.rsppv.criminalintent.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.*
import android.widget.ImageView
import android.widget.TextView
import ru.rsppv.criminalintent.R
import ru.rsppv.criminalintent.model.Crime
import ru.rsppv.criminalintent.model.CrimeLab

class CrimeListFragment : Fragment() {
    private lateinit var mCrimeRecyclerView: RecyclerView
    private lateinit var mNoCrimesView: TextView
    private var mAdapter: CrimeAdapter? = null
    private var mSubtitleVisible = false
    private var mCallbacks: Callbacks? = null

    interface Callbacks {
        fun onCrimeSelected(crime: Crime)
        fun onCrimeSwiped(crime: Crime)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        // should be well documented and possibly throw appropriate exception
        mCallbacks = context as Callbacks
    }

    override fun onDetach() {
        super.onDetach()
        mCallbacks = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater?.inflate(R.menu.fragment_crime_list, menu)
        val subtitleItem = menu?.findItem(R.id.show_subtitle)
        subtitleItem?.setTitle(if (mSubtitleVisible) R.string.hide_subtitle else R.string.show_subtitle)

    }

    override fun onOptionsItemSelected(item: MenuItem?) = when (item?.itemId) {
        R.id.new_crime -> {
            createNewCrime()
            true
        }
        R.id.show_subtitle -> {
            mSubtitleVisible = !mSubtitleVisible
            activity?.invalidateOptionsMenu()
            updateSubtitle()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(SAVED_SUBTITLE_VISIBLE, mSubtitleVisible)
    }

    private fun updateSubtitle() {
        val crimeCount = CrimeLab.getInstance(activity).getAllCrimes().size
        val subtitle =
            if (mSubtitleVisible) getSubtitle(crimeCount) else null
        (activity as AppCompatActivity).supportActionBar?.subtitle = subtitle

    }

    private fun getSubtitle(crimeCount: Int) =
        resources.getQuantityString(R.plurals.subtitle_plurals, crimeCount, crimeCount)

    private fun createNewCrime() {
        activity?.let {
            val newCrime = Crime()
            CrimeLab.getInstance(it).addCrime(newCrime)
            updateUI()
            mCallbacks?.onCrimeSelected(newCrime)
        }
    }


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_crime_list, container, false)
        mCrimeRecyclerView = view.findViewById(R.id.crime_recycler_view)
        mCrimeRecyclerView.layoutManager = LinearLayoutManager(activity)
        ItemTouchHelper(mRevoceOnSwipeHandler).attachToRecyclerView(mCrimeRecyclerView)

        mNoCrimesView = view.findViewById(R.id.no_crimes_text)

        savedInstanceState?.let {
            mSubtitleVisible = it.getBoolean(SAVED_SUBTITLE_VISIBLE)
        }

        return view
    }

    private val mRevoceOnSwipeHandler =
        object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            override fun onMove(
                recyclerView: RecyclerView?,
                viewHolder: RecyclerView.ViewHolder?,
                target: RecyclerView.ViewHolder?
            ): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder?, direction: Int) {
                viewHolder?.let {
                    (mCrimeRecyclerView.adapter as CrimeAdapter).removeAt(it.adapterPosition)
                }
            }
        }

    override fun onResume() {
        super.onResume()
        updateUI()
        updateSubtitle()
    }

    fun updateUI() {
        val crimes = CrimeLab.getInstance(activity).getAllCrimes()
        if (mAdapter == null) {
            mAdapter = CrimeAdapter(crimes)
            mCrimeRecyclerView.adapter = mAdapter
        } else {
            mAdapter?.updateCrimes(crimes)
            mAdapter?.notifyDataSetChanged()
        }
        mNoCrimesView.visibility = if (crimes.isEmpty()) View.VISIBLE else View.GONE

    }

    private inner class CrimeHolder(inflater: LayoutInflater, parent: ViewGroup) :
        RecyclerView.ViewHolder(inflater.inflate(R.layout.list_item_crime, parent, false)),
        View.OnClickListener {
        private val mTitleTextView: TextView
        private val mDateTextView: TextView
        private val mCrimeSolvedImageView: ImageView
        private lateinit var mCrime: Crime

        init {
            itemView.setOnClickListener(this)
            mTitleTextView = itemView.findViewById<View>(R.id.crime_title) as TextView
            mDateTextView = itemView.findViewById<View>(R.id.crime_date) as TextView
            mCrimeSolvedImageView = itemView.findViewById<View>(R.id.crime_solved) as ImageView
        }

        fun bind(crime: Crime) {
            mCrime = crime
            mTitleTextView.text = crime.title
            mDateTextView.text = crime.getDateString(activity)
            mCrimeSolvedImageView.visibility = if (crime.isSolved) View.VISIBLE else View.GONE
        }

        override fun onClick(v: View) {
            mCallbacks?.onCrimeSelected(mCrime)
        }
    }


    private inner class CrimeAdapter(var crimes: MutableList<Crime>) :
        RecyclerView.Adapter<CrimeHolder>() {

        fun updateCrimes(crimeList: MutableList<Crime>) {
            crimes = crimeList
        }

        fun removeAt(position: Int) {
            val crime = crimes[position]
            CrimeLab.getInstance(activity).removeCrime(crime)
            crimes.removeAt(position)
            mCallbacks?.onCrimeSwiped(crime)
            notifyItemRemoved(position)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CrimeHolder {
            return CrimeHolder(LayoutInflater.from(activity), parent)
        }

        override fun onBindViewHolder(holder: CrimeHolder, position: Int) {
            holder.bind(crimes[position])
        }

        override fun getItemCount(): Int {
            return crimes.size
        }
    }

    companion object {
        const val SAVED_SUBTITLE_VISIBLE = "subtitle_visibility"
    }
}
